# Command Line Cheat Sheet
This is a cheat sheet for the codes learned during the [Command Line Exercise](Command-Line_Exercise.md).

>ğŸ’¡**Tip:** You can navigate quickly through this document by using the table of contents in the upper right hand corner:
>![ToC Image](https://docs.github.com/assets/cb-142339/mw-1440/images/help/repository/headings-toc.webp)

## HPC File Structure
As a refresher, this is an example of the HPC file structure: 
```
â””â”€â”€ Zamarin Lab Home Base (/sc/arion/projects/NGSCRC)/
    â”œâ”€â”€ master_data (Raw data, organized by project)/
    â”‚   â”œâ”€â”€ Project 1
    â”‚   â””â”€â”€ Project 2
    â”œâ”€â”€ Scripts (Scripts, organized by project and by pipeline step)/
    â”‚   â”œâ”€â”€ Project 1/
    â”‚   â”‚   â”œâ”€â”€ fastp Script
    â”‚   â”‚   â””â”€â”€ Genome Alignment Script
    â”‚   â””â”€â”€ Project 2/
    â”‚       â”œâ”€â”€ fastp Script
    â”‚       â””â”€â”€ Genome Alignment Script
    â”œâ”€â”€ Work (Output from scripts, organized by project and by pipeline step)/
    â”‚   â”œâ”€â”€ Project 1/
    â”‚   â”‚   â”œâ”€â”€ Trimmed Reads (fastp output)
    â”‚   â”‚   â””â”€â”€ Aligned Files (Genome Alignment output)
    â”‚   â””â”€â”€ Project 2/
    â”‚       â”œâ”€â”€ Trimmed Reads
    â”‚       â””â”€â”€ Aligned Files
    â”œâ”€â”€ Tools (Shared modules for specialty projects not already available on Minerva)/
    â”‚   â”œâ”€â”€ MultiQC
    â”‚   â””â”€â”€ MiXCR
    â””â”€â”€ Resources (Shared resources such as reference genomes)/
        â”œâ”€â”€ Reference Genome (WGS)
        â””â”€â”€ Reference Genome (WES)
```
This structure becomes particularly useful as projects become more complex.

<details>
<summary>Expand for a more detailed example to illustrate this complexity:</summary>
  
### Complex file structures

In this example, we have `Project 1` with two samples (`Sample 1` and `Sample 2`), and `Project 2` with 1 sample (`Sample A`). 

```
.
â””â”€â”€ Zamarin Lab Home Base (/sc/arion/projects/NGSCRC)/
    â”œâ”€â”€ master_data/
    â”‚   â”œâ”€â”€ Project 1/
    â”‚   â”‚   â”œâ”€â”€ Sample 1/
    â”‚   â”‚   â”‚   â”œâ”€â”€ Sample_1_raw_read_1.fasq
    â”‚   â”‚   â”‚   â””â”€â”€ Sample_1_raw_read_2.fasq
    â”‚   â”‚   â””â”€â”€ Sample 2/
    â”‚   â”‚       â”œâ”€â”€ Sample_2_raw_read_1.fasq
    â”‚   â”‚       â””â”€â”€ Sample_2_raw_read_2.fasq
    â”‚   â””â”€â”€ Project 2/
    â”‚       â””â”€â”€ Sample A/
    â”‚           â”œâ”€â”€ Sample_A_raw_read_1.fasq
    â”‚           â””â”€â”€ Sample_A_raw_read_2.fasq
    â”œâ”€â”€ Scripts/
    â”‚   â”œâ”€â”€ Project 1/
    â”‚   â”‚   â”œâ”€â”€ fastp Script/
    â”‚   â”‚   â”‚   â”œâ”€â”€ project_1_sample_list.txt
    â”‚   â”‚   â”‚   â””â”€â”€ project_1_fastp.sh
    â”‚   â”‚   â””â”€â”€ Genome Alignment Script/
    â”‚   â”‚       â”œâ”€â”€ project_1_trimmed_sample_list.txt
    â”‚   â”‚       â””â”€â”€ project_1_align.sh
    â”‚   â””â”€â”€ Project 2/
    â”‚       â”œâ”€â”€ fastp Script/
    â”‚       â”‚   â”œâ”€â”€ project_2_sample_list.txt
    â”‚       â”‚   â””â”€â”€ project_2_fastp.sh
    â”‚       â””â”€â”€ Genome Alignment Script/
    â”‚           â”œâ”€â”€ project_2_trimmed_sample_list.txt
    â”‚           â””â”€â”€ project_2_align.sh
    â”œâ”€â”€ Work/
    â”‚   â”œâ”€â”€ Project 1/
    â”‚   â”‚   â”œâ”€â”€ Trimmed Reads/
    â”‚   â”‚   â”‚   â”œâ”€â”€ Sample 1/
    â”‚   â”‚   â”‚   â”‚   â”œâ”€â”€ Sample_1_trimmed_1.fastq
    â”‚   â”‚   â”‚   â”‚   â””â”€â”€ Sample_1_trimmed_2.fastq
    â”‚   â”‚   â”‚   â””â”€â”€ Sample 2/
    â”‚   â”‚   â”‚       â”œâ”€â”€ Sample_2_trimmed_1.fastq
    â”‚   â”‚   â”‚       â””â”€â”€ Sample_2_trimmed_2.fastq
    â”‚   â”‚   â””â”€â”€ Aligned Files/
    â”‚   â”‚       â”œâ”€â”€ Sample 1/
    â”‚   â”‚       â”‚   â””â”€â”€ Sample_1_aligned.bam
    â”‚   â”‚       â””â”€â”€ Sample 2/
    â”‚   â”‚           â””â”€â”€ Sample_2_aligned.bam
    â”‚   â””â”€â”€ Project 2/
    â”‚       â”œâ”€â”€ Trimmed Reads/
    â”‚       â”‚   â””â”€â”€ Sample A/
    â”‚       â”‚       â”œâ”€â”€ Sample_A_trimmed_1.fastq
    â”‚       â”‚       â””â”€â”€ Sample_A_trimmed_2.fastq
    â”‚       â””â”€â”€ Aligned Files/
    â”‚           â””â”€â”€ Sample A/
    â”‚               â””â”€â”€ Sample_A_aligned.bam
    â”œâ”€â”€ Tools/
    â”‚   â”œâ”€â”€ MultiQC
    â”‚   â””â”€â”€ MiXCR
    â””â”€â”€ Resources/
        â”œâ”€â”€ Reference Genome (WGS)
        â””â”€â”€ Reference Genome (WES)

```
</details>

*File treesğŸŒ²created using [this cool tool](https://tree.nathanfriend.com/)*

Visualizing the structure of your directories can be useful when thinking about navigating the command line.

## ğŸ—ºï¸Navigating the Command Line
These are the codes we learned for basic navigation around the command line environment:

### Change directory: `cd`
Use this code to move from your current directory to another directory.

For example this code brings us to the `master_data` folder where the raw .fastqs for the [Command Line Exercise](Command-Line_Exercise.md) are stored:
```
cd /sc/arion/projects/NGSCRC/master_data/test/Umbrella_Academy
```
If you don't specify a destination (as we did when we included the path to the `master_data` above), typing just `cd` will send you to your login directory.

Let's say you want to move up two folders to the `/sc/arion/projects/NGSCRC/master_data/` folder. 
- You could type out the whole path (`cd /sc/arion/projects/NGSCRC/master_data/`).
- But you can also use "dot dot" (`..`), which is recognized as the directory immediately above the current directory (called the "parent" directory).
- If you know how many folders up you want to travel, you can separate your dot dots with a slash to get to the desired destination.

Since we are moving up two directories, we would type
```
cd ../../
```
### Print working directory: `pwd`
Use this to display the full path of your current directory.

This is useful to verify where you are in your file structures, particularly if you are using `..` and not absolute paths to travel.

In the example above, it would be best to run `cd ../../` followed by `pwd` to verify that you are in the directory you intended to reach. 

As we learned in the exercise when we generated the sample list, `pwd` can also be commbined with other commands (like `printf '%s\n'`, to capture full paths to files).

### List contents: `ls`
Use this to list the contents of your current directory. 


